[16.0][MIG] shopinvader_api_quotation

matthieusaison opened the pull request:
Continuation of this PR: https://github.com/shopinvader/odoo-shopinvader/pull/1447

In this PR we split model and views in sale_quotation module.
And add shopinvader_sale_state in dependency (https://github.com/shopinvader/odoo-shopinvader/pull/1448)

depend on 
- [x] https://github.com/shopinvader/odoo-shopinvader/pull/1466

Commit:
separate the model and view of shopinvader_quotation

Commit:
update sale_quotation and add shopinvader_sale_state

Commit:
migrate shopinvader quotation to fastapi

sebastienbeau commented on shopinvader_quotation/__manifest__.py:
```diff
@@ -12,11 +12,9 @@
     "author": "Akretion",
     "license": "AGPL-3",
-    "depends": ["shopinvader"],
+    "depends": ["sale_quotation"],
```
["shopinvader",  "sale_quotation"]

sebastienbeau commented on sale_quotation/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+    "website": "https://github.com/shopinvader/odoo-shopinvader",
+    "author": "Akretion",
+    "license": "AGPL-3",
+    "depends": ["sale"],
```
"sale_cart"

sebastienbeau commented on sale_quotation/tests/test_notification.py:
```diff
@@ -1,20 +0,0 @@
-# Copyright 2017 Akretion (http://www.akretion.com).
```
should be keep in shopinvader_quotation

sebastienbeau commented on sale_quotation/__manifest__.py:
```diff
@@ -10,7 +10,7 @@
     "author": "Akretion",
     "license": "AGPL-3",
-    "depends": ["sale"],
+    "depends": ["sale", "shopinvader_sale_state"],
```
"sale_cart"

Commit:
add shopinvader_state

sebastienbeau committed:
shopinvader_api_quotation: add possibility to pass params at confirm and add download report

sebastienbeau committed:
sale_quotation: review UI

- change behaviour so you can use it without shopinvader context
- remove typology in favor of a quotation_state
- rework backend UI to make it more human understandable
- adapt test

matthieusaison commented:
:wave: Hello @sbidoul, @lmignon can you take a look and merge this PR please.

lmignon commented on sale_quotation/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1 @@
+This module handle the model and view for product and sale.order needs for shopinvader-api-quotation
```
@matthieusaison Can you expose the functional scope of this addon and how it can be used by an end user. The only way we can figure out how it should work is by reading the code and it's not crystal clear for me. A good way to do it is to explain the 'Why' on a functional point of view.

lmignon replied:
You should define the context and the 'why' of this addon. This description doesn't provide the minimal information to know when to use it.

lmignon commented on shopinvader_api_quotation/routers/cart.py:
```diff
@@ -0,0 +1,34 @@
+def request_quotation(
+    env: Annotated[api.Environment, Depends(authenticated_partner_env)],
+    partner: Annotated["ResPartner", Depends(authenticated_partner)],
+    uuid: str | None = None,
```
```suggestion
    uuid: UUID | None = None,
```

matthieusaison replied:
done

lmignon commented on shopinvader_api_quotation/schemas/sale.py:
```diff
@@ -0,0 +1,27 @@
+        res.customer_ref = odoo_rec.client_order_ref or None
+        # res.shop_only_quotation = any(
+        #     odoo_rec.order_line.product_id.mapped("shop_only_quotation")
+        # ) mettre un champs calculé coté odoo sur model sale_order
```
This lines should be removed

lmignon commented on shopinvader_api_quotation/schemas/sale.py:
```diff
@@ -0,0 +1,27 @@
+        return res
+
+
+class QuotationUpdateInput(StrictExtendableBaseModel):
```
```suggestion
class QuotationUpdateInput(StrictExtendableBaseModel, extra="ignore"):
```

see motivation here https://github.com/shopinvader/odoo-shopinvader/commit/e5331c5906a2a6944076491918d0610ef0a0a9ce

sebastienbeau commented:
@matthieusaison be carefull when you do a rebase, you have drop my commit :'(.
I just re-push them

bealdav commented on shopinvader_api_quotation/routers/quotation.py:
```diff
@@ -0,0 +1,137 @@
+class ShopinvaderApiSaleSalesRouterHelper(models.AbstractModel):
+    _name = "shopinvader_api_quotation.quotations_router.helper"
+    _inherit = "shopinvader_api_sale.sales_router.helper"
+
```
```suggestion
    _description = "Shopinvader api sale router helper"

```

matthieusaison committed:
Update shopinvader_api_quotation/routers/quotation.py

Co-authored-by: David Beal <david.beal@akretion.com>

Commit:
complete readme

sebastienbeau reviewed (approved):
LGTM (code review). Thanks for the rebase

sebastienbeau commented:
@lmignon let's merge it ?

lmignon commented on sale_quotation/models/sale_order.py:
```diff
@@ -0,0 +1,71 @@
+        self.write({"quotation_state": "customer_request", "typology": "sale"})
+        return True
+
+    def _compute_available_for_quotation(self):
+        for record in self:
+            record.available_for_quotation = True
```
@sebastienbeau  What's the purpose of this compute method (and the field behind)?

sebastienbeau replied:
The purpose is to make visible or not on the website if the button "request quotation", but we can remove this compute field and simply "inherit" the Schema sale to return something different

lmignon commented on shopinvader_api_quotation/routers/cart.py:
```diff
@@ -0,0 +1,35 @@
+from odoo.addons.shopinvader_schema_sale.schemas.sale import Sale
+
+
+@cart_router.post("/{uuid}/request_quotation")
```
```suggestion
@cart_router.post("/{uuid}/request_quotation")
@cart_router.post("/current/request_quotation")
@cart_router.post("/request_quotation")
```

lmignon commented on shopinvader_api_quotation/routers/cart.py:
```diff
@@ -0,0 +1,35 @@
+class ShopinvaderApiCartRouterHelper(models.AbstractModel):
+    _inherit = "shopinvader_api_cart.cart_router.helper"
+
+    def _request_quotation(self, partner, uuid):
```
```suggestion
    def _request_quotation(self, partner: ResPartner, uuid: UUID | None=None):
```

lmignon commented on shopinvader_api_quotation/routers/cart.py:
```diff
@@ -0,0 +1,35 @@
+@cart_router.post("/{uuid}/request_quotation")
+def request_quotation(
+    env: Annotated[api.Environment, Depends(authenticated_partner_env)],
+    partner: Annotated["ResPartner", Depends(authenticated_partner)],
```
```suggestion
    partner: Annotated[ResPartner, Depends(authenticated_partner)],
```

lmignon commented on shopinvader_api_quotation/schemas/sale.py:
```diff
@@ -0,0 +1,28 @@
+class Sale(Sale, extends=True):
+    available_for_quotation: bool | None = None
+    shop_only_quotation: bool | None = None
+    customer_ref: str | None = None
```
It's a little strange to introduce the customer ref into this addon since its usage seems to me wider than the quotation context.

sebastienbeau replied:
We can move this it the sale schema and also add the possibility to set it from the cart API

lmignon commented on shopinvader_api_quotation/tests/test_shopinvader_api_quotation.py:
```diff
@@ -0,0 +1,132 @@
+                "quotation_state": "waiting_acceptation",
+            }
+        )
+        with self._create_test_client(router=quotation_router) as test_client:
```
the `quotation_router` router is already defined as the default one into your `setUpClass` method.
```suggestion
        with self._create_test_client() as test_client:
```

lmignon commented on shopinvader_api_quotation/tests/test_shopinvader_api_quotation.py:
```diff
@@ -0,0 +1,132 @@
+                "quotation_state": "waiting_acceptation",
+            }
+        )
+        with self._create_test_client(router=quotation_router) as test_client:
```
```suggestion
        with self._create_test_client() as test_client:
```

lmignon commented on shopinvader_api_quotation/tests/test_shopinvader_api_quotation.py:
```diff
@@ -0,0 +1,132 @@
+                "quotation_state": "waiting_acceptation",
+            }
+        )
+        with self._create_test_client(router=quotation_router) as test_client:
```
```suggestion
        with self._create_test_client() as test_client:
```

lmignon commented on shopinvader_api_quotation/tests/test_shopinvader_api_quotation.py:
```diff
@@ -0,0 +1,132 @@
+        cart = self.env["sale.order"]._create_empty_cart(
+            self.default_fastapi_authenticated_partner.id
+        )
+        with self._create_test_client(router=cart_router) as test_client:
```
```suggestion
        with self._create_test_client() as test_client:
```

lmignon commented on shopinvader_api_quotation/tests/test_shopinvader_api_quotation.py:
```diff
@@ -0,0 +1,132 @@
+                "quotation_state": "waiting_acceptation",
+            }
+        )
+        with self._create_test_client(router=quotation_router) as test_client:
```
```suggestion
        with self._create_test_client() as test_client:
```

lmignon reviewed (changes requested):
Thank you for this proposal. Some comments...

sebastienbeau committed:
sale_quotation/shopinvader_api_quotation: update readme

sebastienbeau commented on shopinvader_api_quotation/routers/quotation.py:
```diff
@@ -0,0 +1,138 @@
+    partner: Annotated[ResPartner, Depends(authenticated_partner)],
+) -> FileResponse:
+    """Download document."""
+    filename, pdf = (
```
TODO use report method merged here : https://github.com/shopinvader/odoo-shopinvader/pull/1489/files#diff-8661a778244e3a221d47f4c85ae9f1e05f275980e48ad3824616e3f10aab4556L118

sebastienbeau committed:
shopinvader_api_quotation: fix typo

Co-authored-by: Laurent Mignon (ACSONE) <laurent.mignon@acsone.eu>

lmignon commented on shopinvader_api_quotation/routers/cart.py:
```diff
@@ -0,0 +1,37 @@
+    authenticated_partner,
+    authenticated_partner_env,
+)
+from odoo.addons.shopinvader_api_cart.routers import cart_router
+from odoo.addons.shopinvader_schema_sale.schemas.sale import Sale
+
+
+@cart_router.post("/{uuid}/request_quotation")
```
To not break extension mechanism from odoo, you must define a new router and explain into your readme that your new router must be mounted at the same point as the initial cart router as I did it for the `shopinvader_api_sale_loyalty` addon in https://github.com/shopinvader/odoo-shopinvader/pull/1558/commits/2c20ea840bdd15190d1466cd96eeee0cb3d81416

paradoxxxzero commented:
Superseded by #1573

sebastienbeau commented:
replaced by : https://github.com/shopinvader/odoo-shopinvader/pull/1573
